### PR TITLE
Ensure proper tracking/cleanup of _bufferToDocumentInCurrentContextMa…

### PR DESCRIPTION
**Pull request by mavasani**
…p when closing shared documents

Current implementation was leaking entries in this map for shared document buffers when the shared document was closed. The primary reason was that we were only tracking the active document ID corresponding to a given text container. Hence, on document close for a specific document ID, we could not determine if we still need to track the text container for other related document IDs or not. This leak in turn was causing stale entries in the map to be returned when the shared document was re-opened.

Fix is to track all the related documents that have been opened with a specific text container, and remove the textContainer tracking entry in _bufferToDocumentInCurrentContextMap when all the attached document IDs have been closed.

Fixes https://github.com/dotnet/roslyn/issues/5343. I verified that this fixes the repro for both shared projects and cross-targeting dotnet core projects.

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet-bot test vsi please

**Comment by mavasani** [comment]
@dpoeschl Can you please do a first round of review before I tag the rest of the IDE team?

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p0_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot test vsi please

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ public abstract partial class Workspace
         private readonly Dictionary<ProjectId, ISet<DocumentId>> _projectToOpenDocumentsMap = new Dictionary<ProjectId, ISet<DocumentId>>();
 
         // text buffer maps
-        private readonly Dictionary<SourceTextContainer, DocumentId> _bufferToDocumentInCurrentContextMap = new Dictionary<SourceTextContainer, DocumentId>();
+        private readonly Dictionary<SourceTextContainer, OneOrMany<DocumentId>> _bufferToDocumentInCurrentContextMap = new Dictionary<SourceTextContainer, OneOrMany<DocumentId>>();
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: It's a little strange for the "_bufferToDocumentInCurrentContextMap" to map to multiple documents, given its name and purpose, and the fact that the 0th element has to be _special_ in a non-obvious way. I was imagining some secondary data structure to track all of the documentIds related to the SourceTextContainer and then removing the SourceTextContainer from the dictionary when all of the documentIds are removed for it.

I won't block merging based on this, but do take a minute to consider this alternate approach. I'd like to know what you think about it.

Comment (mavasani) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: @dpoeschl That actually sounds more cleaner, let me split this into 2 maps to make the active context tracking more explicit.

**Comment by mavasani** [comment]
@dotnet-bot test vsi please

**Comment by mavasani** [comment]
@dpoeschl I have addressed your feedback, thanks!

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-ide for further reviews.

Tagging @srivatsn as we may want to bar check this for 15.1

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p3_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by mavasani** [comment]
Rebased to master as per @srivatsn's recommendation.

**Comment by mavasani** [comment]
@dotnet-bot test vsi please

**Comment by mavasani** [comment]
@dotnet-bot test windows_eta_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot test windows_eta_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot test windows_vsi_p2_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
Same as https://github.com/dotnet/roslyn/pull/17424#issuecomment-283752429 - failures are unrelated and all VSI tests passed prior to rebasing from d15.1.x to master. I am going to merge this in unless there are any concerns.

**Comment by dpoeschl** [comment]
No concerns from me w.r.t. the "failing" tests.

**Pull request merged**